File: 2016/day-1/part-2/src/main.rs
enum Directions {
    North,
    East,
    South,
    West,
}

struct point {
    x: i32,
    y: i32,
    direction: Directions,
}

fn main() {
    let mut position = point {
        x: 0,
        y: 0,
        direction: Directions::North,
    };
    let instructions = std::fs::read_to_string("input.txt").expect("Error reading input.txt");

    // Find the first location visited twice
    let mut visited = std::collections::HashSet::new();
    visited.insert((0, 0));
    for token in instructions.split(",").map(|s| s.trim()) {
        let steps: i32 = token[1..].parse().expect("Error parsing steps");
        match token.chars().nth(0).unwrap() {
            'R' => match position.direction {
                Directions::North => position.direction = Directions::East,
                Directions::East => position.direction = Directions::South,
                Directions::South => position.direction = Directions::West,
                Directions::West => position.direction = Directions::North,
            },
            'L' => match position.direction {
                Directions::North => position.direction = Directions::West,
                Directions::East => position.direction = Directions::North,
                Directions::South => position.direction = Directions::East,
                Directions::West => position.direction = Directions::South,
            },
            _ => panic!("Invalid instruction"),
        }
        match position.direction {
            Directions::North => {
                for i in 1..=steps {
                    if visited.contains(&(position.x, position.y + i)) {
                        println!("Distance: {}", position.x.abs() + (position.y + i).abs());
                        return;
                    }
                    visited.insert((position.x, position.y + i));
                }
                position.y += steps;
            }
            Directions::East => {
                for i in 1..=steps {
                    if visited.contains(&(position.x + i, position.y)) {
                        println!("Distance: {}", (position.x + i).abs() + position.y.abs());
                        return;
                    }
                    visited.insert((position.x + i, position.y));
                }
                position.x += steps;
            }
            Directions::South => {
                for i in 1..=steps {
                    if visited.contains(&(position.x, position.y - i)) {
                        println!("Distance: {}", position.x.abs() + (position.y - i).abs());
                        return;
                    }
                    visited.insert((position.x, position.y - i));
                }
                position.y -= steps;
            }
            Directions::West => {
                for i in 1..=steps {
                    if visited.contains(&(position.x - i, position.y)) {
                        println!("Distance: {}", (position.x - i).abs() + position.y.abs());
                        return;
                    }
                    visited.insert((position.x - i, position.y));
                }
                position.x -= steps;
            }
        }
    }
}

File: 2016/day-1/part-2/main.cpp
#include <boost/algorithm/string.hpp>
#include <fstream>
#include <iostream>
#include <set>
#include <sstream>
#include <string>
#include <vector>

auto main() -> int {
  std::ifstream file("input.txt");
  std::string instruction = "";
  std::vector<std::string> instructions = {};
  std::string pause = "";
  int x = 0;
  int y = 0;
  char heading = 'N';
  int steps = 0;
  std::string direction = "";
  std::set<std::pair<int, int>> visited = {};
  bool found = false;

  while (std::getline(file, instruction, ',')) {
    boost::trim(instruction);
    instructions.push_back(instruction);
  }

  for (auto i : instructions) {
    direction = i.substr(0, 1);
    steps = std::stoi(i.substr(1, i.length()));

    if (direction == "R") {
      if (heading == 'N') {
        heading = 'E';
        for (int j = 0; j < steps; j++) {
          x++;
          if (visited.find(std::make_pair(x, y)) != visited.end()) {
            found = true;
            break;
          }
          visited.insert(std::make_pair(x, y));
        }
      } else if (heading == 'E') {
        heading = 'S';
        for (int j = 0; j < steps; j++) {
          y--;
          if (visited.find(std::make_pair(x, y)) != visited.end()) {
            found = true;
            break;
          }
          visited.insert(std::make_pair(x, y));
        }
      } else if (heading == 'S') {
        heading = 'W';
        for (int j = 0; j < steps; j++) {
          x--;
          if (visited.find(std::make_pair(x, y)) != visited.end()) {
            found = true;
            break;
          }
          visited.insert(std::make_pair(x, y));
        }
      } else if (heading == 'W') {
        heading = 'N';
        for (int j = 0; j < steps; j++) {
          y++;
          if (visited.find(std::make_pair(x, y)) != visited.end()) {
            found = true;
            break;
          }
          visited.insert(std::make_pair(x, y));
        }
      }
    } else if (direction == "L") {
      if (heading == 'N') {
        heading = 'W';
        for (int j = 0; j < steps; j++) {
          x--;
          if (visited.find(std::make_pair(x, y)) != visited.end()) {
            found = true;
            break;
          }
          visited.insert(std::make_pair(x, y));
        }
      } else if (heading == 'W') {
        heading = 'S';
        for (int j = 0; j < steps; j++) {
          y--;
          if (visited.find(std::make_pair(x, y)) != visited.end()) {
            found = true;
          }
          visited.insert(std::make_pair(x, y));
        }
      } else if (heading == 'S') {
        heading = 'E';
        for (int j = 0; j < steps; j++) {
          x++;
          if (visited.find(std::make_pair(x, y)) != visited.end()) {
            found = true;
            break;
          }
          visited.insert(std::make_pair(x, y));
        }
      } else if (heading == 'E') {
        heading = 'N';
        for (int j = 0; j < steps; j++) {
          y++;
          if (visited.find(std::make_pair(x, y)) != visited.end()) {
            found = true;
            break;
          }
          visited.insert(std::make_pair(x, y));
        }
      }
    }
  }

  std::cout << "The first location visited twice is " << x << ", " << y
            << " which is " << abs(x) + abs(y) << " blocks away." << std::endl;
  return 0;
}